libraries/SPI/SPI.cpp

```diff
@@ -311,53 +319,55 @@ void SPIClass::write32(uint32_t data, bool msb) {
 
 /**
  * Note:
- *  data need to be aligned to 32Bit
- *  or you get an Fatal exception (9)
+ *  data needs to be aligned to 32 bit boundaries like this: char outBuffer[spiBufferSize] __attribute__ ((aligned (4)));
+ *  if not you may get an Fatal exception (9)
  * @param data uint8_t *
  * @param size uint32_t
  */
 void SPIClass::writeBytes(uint8_t * data, uint32_t size) {
+	uint32_t localSize;
     while(size) {
-        if(size > 64) {
-            writeBytes_(data, 64);
-            size -= 64;
-            data += 64;
-        } else {
-            writeBytes_(data, size);
-            size = 0;
-        }
-    }
+        if(size > 64)
+			localSize = 64;
+		else
+			localSize = size;
+			
+        writeBytes_(data, localSize);
+        size -= localSize;
+		if ((size != 0) && (data)) data += localSize;
+	}
 }
 
-void SPIClass::writeBytes_(uint8_t * data, uint8_t size) {
+
+inline void SPIClass::writeBytes_(uint8_t * data, uint32_t size) {
+#ifndef SPI_SKIP_INITIAL_BUSY_CHECK
     while(SPI1CMD & SPIBUSY) {}
-    // Set Bits to transfer
+#endif
+    // Set bits to transfer
     setDataBits(size * 8);
 
-    volatile uint32_t * fifoPtr = &SPI1W0;
-    uint32_t * dataPtr = (uint32_t*) data;
-    uint8_t dataSize = ((size + 3) / 4);
-
-    while(dataSize--) {
-        *fifoPtr = *dataPtr;
-        dataPtr++;
-        fifoPtr++;
-    }
-
+	if(data) {
+		memcpy((uint8_t*)&SPI1W0, data, size);
```

Comment:
Memcpy will always use load/store instructions with correct alignment, so unaligned load/store exceptions are not the issue here. The issue may be that memcpy will try to write a smaller than 4-byte value into the peripheral register (i.e. perform a 2-byte or 1-byte write, aligned on 2-byte or 1-byte boundary, respectively). Doing so will not cause any exception, but the SPI FIFO was not designed to support 2-byte or 1-byte writes. So you may possibly end up with incorrect data being sent.
I'll try to come up with some test case which illustrates this.